generate_index_html.py

```diff
@@ -0,0 +1,148 @@
+    version_to_files = {}
+    # Example -
+    # version_to_files = {
+    #   "2.0.2": (
+    #       "46.25 KiB",
+    #       "2022-07-20T21:18:06Z",
+    #       "apache-airflow-providers-cncf-kubernetes/apache_airflow_providers_cncf_kubernetes-1.0.2rc1-py3-none-any.whl"
+    #   )
+    # }
+    for match in matches:
+        if match[6] not in version_to_files:
+            version_to_files[match[6]] = []
+        version_to_files[match[6]].append((match[0], match[1], match[4]))
+    return version_to_files
+
+
+def get_released_public_package_list(package: str) -> dict:
+    url = f"https://pypi.org/pypi/{package}/json"
+    response = requests.get(url)
+    response.raise_for_status()
+    return {release: files for release, files in response.json()["releases"].items()}
+
+
+def get_custom_package_json(package_info: tuple) -> dict:
+    """Generate dict data for a custom package"""
+    return {
+        "size": str(package_info[0]).strip(),
+        "upload_time_iso_8601": package_info[1],
+        "url": package_info[2],
+        "filename": package_info[2],
+    }
+
+
+def prepare_index_html(releases: dict, custom_packages: dict, requirements):
+    newline = "\n"
+
+    # random_links will hold the following data:
+    # [
+    #   ('8.3.1', ['0.001 KiB  2024-07-12T20:20:11Z  <a href="...">...cncf_kubernetes-8.3.1-py3-none-any.whl</a>']),
+    #   ('8.3.3', ['0.001 KiB  2024-07-12T20:20:11Z  <a href="...">...cncf_kubernetes-8.3.3-py3-none-any.whl</a>'])
+    # ]
+    random_links = []
+    for version, packages in releases.items():
+        package_version = Version(version)
+        if version not in custom_packages and version not in requirements and not package_version.is_prerelease:
+            random_links.append((package_version, [prepare_package_link(version, package) for package in packages]))
+        elif version in custom_packages:
+            random_links.append((package_version, [prepare_package_link(version, get_custom_package_json(file)) for file in custom_packages[version]]))
+
+    # There could be versions that are not published on the public pip index, but we still want to include them.
+    # Example - `7.0.0+astro.2`
+    remaining_packages_versions = set(custom_packages.keys()) - set(releases)
+    for version in remaining_packages_versions:
+        random_links.append((Version(version), [prepare_package_link(version, get_custom_package_json(file)) for file in custom_packages[version]]))
+
+    # Sort by package version before rendering
+    random_links = sorted(random_links, key=lambda d: d[0])
+    sorted_links = []
+    for link in random_links:
+        sorted_links.extend(link[1])
+
+    return (
+        f"<!DOCTYPE html>\n"
+        f"<html><head><title>Astronomer Python Packages</title></head>\n"
+        f"<body><pre>\n"
+        f"{newline.join([link for link in sorted_links])}"
+        f"\nTOTAL: {len(releases)} objects\n"
+        f"</pre></body></html>"
```

Comment:
```suggestion
        "<!DOCTYPE html>\n"
        "<html><head><title>Astronomer Python Packages</title></head>\n"
        "<body><pre>\n"
        f"{newline.join([link for link in sorted_links])}\n"
        "TOTAL: {len(releases)} objects\n"
        "</pre></body></html>"
```

Don't fstring everything. Probably clearer to refactor using `"""` too.